packages/affiliates/libs/models.js

```diff
@@ -0,0 +1,184 @@
+const assert = require("assert");
+const lodash = require("lodash");
+
+// assumes sorted oldest to newest
+// always return price equal to or earlier than requested time
+// prices are input directly from coingecko api
+// [
+//   [ timestamp, price],
+//   [ timestamp, price],
+// ]
+function Prices(prices = []) {
+  function lookup(time) {
+    const index = lodash.sortedIndexBy(prices, [time], ([time]) => {
+      return time;
+    });
+    if (prices[index] === time) return prices[index];
+    return prices[index - 1];
+  }
+
+  return {
+    lookup,
+    prices
+  };
+}
+
+// users show current ratio of affiliates tagged on thier deposits
+// keyed by user address, and show total attributed for each affiliate per user
+function SharedAttributions() {
+  const addresses = new Map();
+  function create(user) {
+    assert(!addresses.has(user), "Already has user");
+    return set(user, {});
+  }
+  function get(user) {
+    assert(addresses.has(user), "User does not exist");
+    return addresses.get(user);
+  }
+  function getOrCreate(user) {
+    try {
+      return get(user);
+    } catch (err) {
+      return create(user);
+    }
+  }
+  function set(user, data) {
+    addresses.set(user, data);
+    return data;
+  }
+  function attribute(user, affiliate, amount) {
+    assert(user, "requires user");
+    assert(affiliate, "requires affiliate");
+    assert(amount, "requires amount");
+    const data = getOrCreate(user);
+    if (data[affiliate] == null) data[affiliate] = "0";
+    data[affiliate] = (BigInt(data[affiliate]) + BigInt(amount)).toString();
+    return set(user, data);
+  }
+  function calculateShare(user, affiliate, scale = 1000000n) {
```

Comment:
Is scale being used to do fixed point math here?

Reply:
i think so. its to be able to have a usable number which is a fraction of 1.